### PR TITLE
Install pretty on ipython

### PR DESCRIPTION
**Pull request by nathanrpage97**
## Type of changes

- [ ] Bug fix
- [x] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [ ] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

#471  describes the issue of injecting the display hook into ipython. 

The issue is that the sys.displayhook cannot be changed inside of the ipython repl. Instead, I've added the display_hook into the [ipython plain text formatter](https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.formatters.html).

### CODE REVIEW

File: rich/pretty.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def install(
 
     console = console or get_console()
 
-    def display_hook(value: Any) -> None:
+    def display_hook(value: Any, *_args, **_kwargs) -> None:
```
Comment (willmcgugan) [inline rich/pretty.py:66]: Is this for iPython's benefit? sys.displayhook doesn't take any additional argument according to docs.

File: rich/pretty.py [hunk 2]
```diff
@@ -83,7 +83,16 @@ def display_hook(value: Any) -> None:
...
-    sys.displayhook = display_hook
+    try:  # pragma: no cover
+        ip = get_ipython()  # type: ignore
+        from IPython.core.formatters import BaseFormatter
+
+        # replace plain text formatter with rich formatter
+        formatter = BaseFormatter()
+        ip.display_formatter.formatters["text/plain"] = formatter
+        formatter.for_type(object, func=display_hook)
+    except:
```
Comment (willmcgugan) [inline rich/pretty.py:94]: This should probably be `except Exception` so as not to catch any system exceptions.

**Review by willmcgugan — COMMENTED** [review]
This is great. iPython support is a common request.

**Comment by willmcgugan** [comment]
@nathanrpage97 There is an unfortunate glitch where iPython print "In [X]:" and Rich expects the cursor to be at the start of the line. This can break rendering of the first line. It's really obvious for something like Panel, but occurs in other places.

Screen Shot 2020-12-28 at 17 35 42

The default pretty printing in iPython will insert a new line when printing containers like dicts or lists. I'm thinking that Rich should do the same, and also print a new line for Rich renderables.

I think the logic should be something like the following:

```python
requires_newline = (ConsoleRenderable, abc.Mapping, abc.Sequence, abc.Set)
if isinstance(value, requires_newline):
    console.line()
```

What do you think?

File: rich/pretty.py (hunk 1 above)
Comment (nathanrpage97) [reply rich/pretty.py]: It was there to allow it to work with the ipython display hook, but is no longer needed

**Comment by nathanrpage97** [comment]
Alright, I've added in the change you suggested for the new line. Still needs testing in a few environments including jupyter.

**Comment by nathanrpage97** [comment]
Note there seems to be a recursion error issue when running it in jupyter.

**Comment by nathanrpage97** [comment]
Alright, I've tweaked it to make sure rich will play well with terminal ipython and rich display jupyter.

**Comment by willmcgugan** [comment]
Great, thanks.

**Pull request merged**